README.md
![Bluejay](https://raw.githubusercontent.com/steamclock/bluejay/master/bluejay-wordmark.png)

![CocoaPods Compatible](https://img.shields.io/cocoapods/v/Bluejay.svg)
![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)
[![Platform](https://img.shields.io/cocoapods/p/Bluejay.svg?style=flat)](http://cocoadocs.org/docsets/Bluejay)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg)]()


Bluejay is a simple Swift framework for building reliable Bluetooth LE apps.

Bluejay's primary goals are:
- Simplify talking to a single Bluetooth LE peripheral
- Make it easier to handle Bluetooth operations reliably
- Take advantage of Swift features and conventions

## Index

- [Features](#features)
- [Requirements](#requirements)
- [Installation](#installation)
- [Demo](#demo)
- [Usage](#usage)
  - [Initialization](#initialization)
  - [Bluetooth Events](#bluetooth-events)
  - [Services and Characteristics](#services-and-characteristics)
  - [Scanning](#scanning)
  - [Connecting](#connecting)
  - [Disconnect](#disconnect)
  - [Cancel Everything](#cancel-everything)
  - [Auto Reconnect](#auto-reconnect)
  - [Disconnect Handler](#disconnect-handler)
  - [Connection States](#connection-states)
- [Deserialization and Serialization](#deserialization-and-serialization)
  - [Receivable](#receivable)
  - [Sendable](#sendable)
- [Interactions](#interactions)
  - [Reading](#reading)
  - [Writing](#writing)
  - [Listening](#listening)
  - [Background Task](#background-task)
- [Background Restoration](#background-restoration)
  - [Background Permission](#background-permission)
  - [State Restoration](#state-restoration)
  - [Listen Restoration](#listen-restoration)
- [Advanced Usage](#advanced-usage)
  - [Write and Assemble](#write-and-assemble)
  - [Flush Listen](#flush-listen)
  - [CoreBluetooth Migration](#corebluetooth-migration)
  - [Monitor Peripheral Services](#monitor-peripheral-services)

## Features

- A callback-based API
- A FIFO operation queue for more synchronous and predictable behaviour
- A background task mode for batch operations that avoids the "callback pyramid of death"
- Simple protocols for data serialization and deserialization
- An easy and safe way to observe connection states
- Powerful background restoration support
- Extended error handling and logging support

## Requirements

- iOS 11 or later recommended
- Xcode 11.3.1 or later recommended
- Swift 5 or later recommended

## Installation

### CocoaPods

`pod 'Bluejay', '~> 0.8'`

Or to try the latest master:

`pod 'Bluejay', :git => 'https://github.com/steamclock/bluejay.git', :branch => 'master'`

### Carthage

```
github "steamclock/bluejay" ~> 0.8
github "DaveWoodCom/XCGLogger" ~> 6.1.0
```

Refer to [official Carthage documentation](https://github.com/Carthage/Carthage#supporting-carthage-for-your-framework) for the rest of the instructions.

**Note:** `Bluejay.framework`, `ObjcExceptionBridging.framework`, and `XCGLogger.framework` are all required.

### Import

```swift
import Bluejay
```

## Demo

The iOS Simulator does not simulate Bluetooth, and you may not have a debuggable Bluetooth LE peripheral handy, so we have prepared you a pair of demo apps to test with.

1. **BluejayHeartSensorDemo:** an app that can connect to a Bluetooth LE heart sensor.
2. **DittojayHeartSensorDemo:** a virtual Bluetooth LE heart sensor.

#### To try out Bluejay:

1. Get two iOS devices – one to run **Bluejay Demo**, and the other to run **Dittojay Demo**.
2. Grant permission for notifications on **Bluejay Demo**.
3. Grant permission for background mode on **Dittojay Demo**.
4. Connect using **Bluejay Demo**.

#### To try out background restoration (after connecting):

1. In **Bluejay Demo**, tap on "End listen to heart rate".
- This is to prevent the continuous heart rate notification from triggering state restoration right after we terminate the app, as it's much clearer and easier to verify state restoration when we can manually trigger a Bluetooth event at our own leisure and timing.
2. Tap on "Terminate app".
- This will crash the app, but also simulate app termination due to memory pressure, **and** allow CoreBluetooth to cache the current session and wait for Bluetooth events to begin state restoration.
3. In **Dittojay Demo**, tap on "Chirp" to *revive* **Bluejay Demo**
- This will send a Bluetooth event to the device with the terminated **Bluejay Demo**, and its CoreBluetooth stack will wake up the app in the background and execute a few quick tasks, such as scheduling a few local notifications for verification and debugging purposes in this case.

## Usage

### Initialization

To create an instance of Bluejay:

```swift
let bluejay = Bluejay()
```

While it is convenient to create one Bluejay instance and use it everywhere, you can also create instances in specific portions of your app and tear them down after use. It's worth noting, however, that each instance of Bluejay has its own [CBCentralManager](https://developer.apple.com/documentation/corebluetooth/cbcentralmanager), which makes the multi-instance approach somewhat more complex.

Once you've created an instance, you can start running Bluejay, which will then initialize the [CoreBluetooth](https://developer.apple.com/documentation/corebluetooth) session. Note that **instantiating a Bluejay instance and running a Bluejay instance are two separate operations.**

You must always start Bluejay in your AppDelegate's `application(_:didFinishLaunchingWithOptions:)` if you want to support [background restoration](#background-restoration), otherwise you are free to start Bluejay anywhere appropriate in your app. For example, apps that don't require background restoration often initialize and start their Bluejay instance from the initial view controller.

```swift
bluejay.start()
```

If your app needs Bluetooth to work in the background, then you have to support background restoration in your app. While Bluejay has already simplified much of background restoration for you, [it will still take some extra work](#background-restoration), and we also recommend reviewing the [relevant Apple docs](https://developer.apple.com/library/archive/documentation/NetworkingInternetWeb/Conceptual/CoreBluetooth_concepts/CoreBluetoothBackgroundProcessingForIOSApps/PerformingTasksWhileYourAppIsInTheBackground.html). Background restoration is tricky and difficult to get right.

Bluejay also supports [CoreBluetooth migration](#corebluetooth-migration) for working with other Bluetooth libraries or with your own Bluetooth code.

### Bluetooth Events

The `ConnectionObserver` protocol allows a class to monitor and to respond to major Bluetooth and connection-related events:

```swift
public protocol ConnectionObserver: class {
    func bluetoothAvailable(_ available: Bool)
    func connected(to peripheral: PeripheralIdentifier)
    func disconnected(from peripheral: PeripheralIdentifier)
}
```

You can register a connection observer using:

```swift
bluejay.register(connectionObserver: batteryLabel)
```

Unregistering a connection observer is not necessary, because Bluejay only holds weak references to registered observers, so Bluejay will clear nil observers from its list when they are found at the next event's firing. But if you need to do so before that happens, you can use:

```swift
bluejay.unregister(connectionObserver: rssiLabel)
```

### Services and Characteristics

In Bluetooth parlance, a Service is a group of attributes, and a Characteristic is an attribute belonging to a group. For example, BLE peripherals that can detect heart rates typically have a Service named "Heart Rate" with a UUID of "180D". Inside that Service are Characteristics such as "Body Sensor Location" with a UUID of "2A38", as well as "Heart Rate Measurement" with a UUID of "2A37".

Many of these Services and Characteristics are standards specified by the Bluetooth SIG organization, and most hardware adopt their specifications. For example, most BLE peripherals implement the Service "Device Information" which has a UUID of "180A", which is where Characteristics such as firmware version, serial number, and other hardware details can be found. Of course, there are many BLE uses not covered by the Bluetooth Core Spec, and custom hardware often have their own unique Services and Characteristics.

Here is how you can specify Services and Characteristics for use in Bluejay:

```swift
let heartRateService = ServiceIdentifier(uuid: "180D")
let bodySensorLocation = CharacteristicIdentifier(uuid: "2A38", service: heartRateService)
let heartRate = CharacteristicIdentifier(uuid: "2A37", service: heartRateService)
```

Bluejay uses the `ServiceIdentifier` and `CharacteristicIdentifier` structs to avoid problems like accidentally specifying a Service when a Characteristic is expected.

### Scanning

Bluejay has a powerful scanning API that can be be used simply or customized to satisfy many use cases.

CoreBluetooth scans for devices using services. In other words, CoreBluetooth, and therefore Bluejay, expects you to know beforehand one or several public services the peripherals you want to scan for contains.

#### Basic Scanning

This simple call will just notify you when there is a new discovery, and when the scan has finished:

```swift
bluejay.scan(
    serviceIdentifiers: [heartRateService],
    discovery: { [weak self] (discovery, discoveries) -> ScanAction in
        guard let weakSelf = self else {
            return .stop
        }

        weakSelf.discoveries = discoveries
        weakSelf.tableView.reloadData()

        return .continue
    },
    stopped: { (discoveries, error) in
        if let error = error {
            debugPrint("Scan stopped with error: \(error.localizedDescription)")
        }
        else {
            debugPrint("Scan stopped without error.")
        }
})
```

A scan result `(ScanDiscovery, [ScanDiscovery])` contains the current discovery followed by an array of all the discoveries made so far.

The stopped result contains a final list of discoveries available just before stopping, and an error if there is one. If there isn't an error, that means that the scan was stopped intentionally or expectedly.

#### Scan Action

A `ScanAction` is returned at the end of a discovery callback to tell Bluejay whether to keep scanning or to stop.

```swift
public enum ScanAction {
    case `continue`
    case blacklist
    case stop
    case connect(ScanDiscovery, (ConnectionResult) -> Void)
}
```

Returning `blacklist` will ignore any future discovery of the same peripheral within the current scan session. This is only useful when `allowDuplicates` is set to true. See [Apple docs on CBCentralManagerScanOptionAllowDuplicatesKey](https://developer.apple.com/documentation/corebluetooth/cbcentralmanagerscanoptionallowduplicateskey?language=objc) for more info.

Returning `connect` will make Bluejay stop the scan as well as perform your connection request. This is useful if you want to connect right away when you've found the peripheral you're looking for.

**Tip:** You can set up the `ConnectionResult` block outside the scan call to reduce callback nesting.

#### Monitoring

Another useful way to use the scanning API is to scan continuously, i.e. to monitor, for purposes such as observing the RSSI changes of nearby peripherals to estimate their proximity:

```swift
bluejay.scan(
    duration: 15,
    allowDuplicates: true,
    serviceIdentifiers: nil,
    discovery: { [weak self] (discovery, discoveries) -> ScanAction in
        guard let weakSelf = self else {
            return .stop
        }

        weakSelf.discoveries = discoveries
        weakSelf.tableView.reloadData()

        return .continue
    },
    expired: { [weak self] (lostDiscovery, discoveries) -> ScanAction in
        guard let weakSelf = self else {
            return .stop
        }

        debugPrint("Lost discovery: \(lostDiscovery)")

        weakSelf.discoveries = discoveries
        weakSelf.tableView.reloadData()

        return .continue
}) { (discoveries, error) in
        if let error = error {
            debugPrint("Scan stopped with error: \(error.localizedDescription)")
        }
        else {
            debugPrint("Scan stopped without error.")
        }
}
```

Setting `allowDuplicates` to true will stop coalescing multiple discoveries of the same peripheral into one single discovery callback. Instead, you'll get a discovery call every time a peripheral's advertising packet is picked up. This will **consume more battery, and does not work in the background**.

**Warning:** An allow duplicates scan will stop with an error if your app is backgrounded during the scan.

The `expired` callback is only invoked when `allowDuplicates` is true. This is called when Bluejay estimates that a previously discovered peripheral is likely out of range or no longer broadcasting. Essentially, when `allowDuplicates` is set to true, every time a peripheral is discovered a timer associated with that peripheral starts counting down. If that peripheral is within range, and even if it has a slow broadcasting interval, it is likely that peripheral will be picked up by the scan again and cause the timer to refresh. If not and the timer expires without being refreshed, Bluejay makes an educated guess and suggests that the peripheral is no longer reachable. Be aware that this is an estimation.

**Warning**: Setting `serviceIdentifiers` to `nil` will result in picking up all available Bluetooth peripherals in the vicinity, **but is not recommended by Apple**. It may cause **battery and cpu issues** on prolonged scanning, and it also **doesn't work in the background**. It is not a private API call, but an available option where you need a quick solution when testing and prototyping.

**Tip:** Specifying at least one specific service identifier is the most common way to scan for Bluetooth devices in iOS. If you need to scan for all Bluetooth devices, we recommend making use of the `duration` parameter to stop the scan after 5 ~ 10 seconds to avoid scanning indefinitely and overloading the hardware.

### Connecting

It is important to keep in mind that Bluejay is designed to work with a single BLE peripheral. Multiple connections at once is not currently supported, and a connection request will fail if Bluejay is already connected or is still connecting. Although this can be a limitation for some sophisticated apps, it is more commonly a safeguard to ensure your app does not issue connections unnecessarily or erroneously.

```swift
bluejay.connect(selectedSensor, timeout: .seconds(15)) { result in
    switch result {
    case .success:
        debugPrint("Connection attempt to: \(selectedSensor.description) is successful")
    case .failure(let error):
        debugPrint("Failed to connect with error: \(error.localizedDescription)")
    }
}
```

#### Timeouts

You can also specify a timeout for a connection request, default is no timeout:

```swift
public enum Timeout {
    case seconds(TimeInterval)
    case none
}
```

**Tip:** We recommend always setting at least a 15 seconds timeout for your connection requests.

### Disconnect

To disconnect:

```swift
bluejay.disconnect()
```

Bluejay also supports finer controls over your disconnection:

#### Queued Disconnect

A queued disconnect will be queued like all other Bluejay API requests, so the disconnect attempt will wait for its turn until all the queued tasks are finished.

To perform a queued disconnect, simply call:

```swift
bluejay.disconnect()
```

#### Immediate Disconnect

An immediate disconnect will immediately fail and empty all tasks from the queue even if they are still running and then immediately disconnect.

There are two ways to perform an immediate disconnect:

```swift
bluejay.disconnect(immediate: true)
```

```swift
bluejay.cancelEverything()
```

#### Expected vs Unexpected Disconnection

Bluejay's log will describe in detail whether a disconnection is expected or unexpected. This is important when debugging a disconnect-related issue, as well as explaining why Bluejay is or isn't attempting to auto reconnect.

Any explicit call to `disconnect` or `cancelEverything` with disconnect will result in an expected disconnection.

All other disconnection events will be considered unexpected. For examples:
- If a connection attempt fails due to hardware errors and not from a timeout
- If a connected device moves out of range
- If a connected device runs out of battery or is shut off
- If a connected device's Bluetooth module crashes and is no longer negotiable

### Cancel Everything

The reason why there is a `cancelEverything` API in addition to `disconnect`, is because sometimes we want to cancel everything in the queue but **remain** connected.

```swift
bluejay.cancelEverything(shouldDisconnect: false)
```

### Auto Reconnect

By default, `shouldAutoReconnect` is `true` and Bluejay will always try to automatically reconnect after an unexpected disconnection.

Bluejay will only set `shouldAutoReconnect` to `false` under these circumstances:

1. If you manually call `disconnect` and the disconnection is successful.
2. If you manually call `cancelEverything` and its disconnection is successful.

Bluejay will also **always** reset `shouldAutoReconnect` to `true` on a successful connection to a peripheral, as we usually want to reconnect to the same device as soon as possible if a connection is lost unexpectedly during normal usage.

However, there are some cases where auto reconnect is not desirable. In those cases, use a `DisconnectHandler` to evaluate and to override auto reconnect.

### Disconnect Handler

A disconnect handler is a single delegate that is suitable for performing major recovery, retry, or reset operations, such as restarting a scan when there is a disconnection.

The purpose of this handler is to help avoid writing and repeating major resuscitation and error handling logic inside the error callbacks of your regular connect, disconnect, read, write, and listen calls. Use the disconnect handler to perform one-time and significant operations at the very end of a disconnection.

In addition to helping you avoid redundant and conflicted logic in various callbacks when there is a disconnection, the disconnect handler also allows you to evaluate and to control Bluejay's auto-reconnect behaviour.

For example, this protocol implementation will always turn off auto reconnect whenever there is a disconnection, expected or not.

```swift
func didDisconnect(
  from peripheral: PeripheralIdentifier,
  with error: Error?,
  willReconnect autoReconnect: Bool) -> AutoReconnectMode {
    return .change(shouldAutoReconnect: false)
}
```

We also anticipate that for most apps, different view controllers may want to handle disconnection differently, so simply register and replace the existing disconnect handler as your user navigates to different parts of your app.

```swift
bluejay.registerDisconnectHandler(handler: self)
```

Similar to connection observers, you do not have to explicitly unregister unless you need to.

### Connection States

Your Bluejay instance has these properties to help you make connection-related decisions:

- `isBluetoothAvailable`
- `isBluetoothStateUpdateImminent`
- `isConnecting`
- `isConnected`
- `isDisconnecting`
- `shouldAutoReconnect`
- `isScanning`
- `hasStarted`
- `defaultWarningOptions`
- `isBackgroundRestorationEnabled`

## Deserialization and Serialization

Reading, writing, and listening to Characteristics is straightforward in Bluejay. Most of the work involved is building out the deserialization and serialization for your data. Let's have a quick look at how Bluejay helps standardize this process in your app via the `Receivable` and `Sendable` protocols.

#### Receivable

Models that represent data you wish to read and receive from your peripheral should all conform to the `Receivable` protocol.

Here is a partial example for the [Heart Rate Measurement Characteristic](https://www.bluetooth.com/specifications/gatt/viewer?attributeXmlFile=org.bluetooth.characteristic.heart_rate_measurement.xml):

```swift
import Bluejay
import Foundation

struct HeartRateMeasurement: Receivable {

    private var flags: UInt8 = 0
    private var measurement8bits: UInt8 = 0
    private var measurement16bits: UInt16 = 0
    private var energyExpended: UInt16 = 0
    private var rrInterval: UInt16 = 0

    private var isMeasurementIn8bits = true

    var measurement: Int {
        return isMeasurementIn8bits ? Int(measurement8bits) : Int(measurement16bits)
    }

    init(bluetoothData: Data) throws {
        flags = try bluetoothData.extract(start: 0, length: 1)

        isMeasurementIn8bits = (flags & 0b00000001) == 0b00000000

        if isMeasurementIn8bits {
            measurement8bits = try bluetoothData.extract(start: 1, length: 1)
        } else {
            measurement16bits = try bluetoothData.extract(start: 1, length: 2)
        }
    }

}

```

Note how you can use the `extract` function that Bluejay adds to `Data` to easily parse the bytes you need. We have plans to build more protection and error handling for this in the future.

Finally, while it is not essential and it will depend on the context, we suggest only exposing the needed and computed properties of your models.

#### Sendable

Models representing data you wish to send to your peripheral should all conform to the `Sendable` protocol. In a nutshell, this is how you help Bluejay determine how to convert your models into `Data`:

```swift
import Foundation
import Bluejay

struct Coffee: Sendable {

    let data: UInt8

    init(coffee: CoffeeEnum) {

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed

## [0.8.9] - 2021-11-23
### Added
- Add setting to broadcast errors sent to cancelEverything to all listeners

### Fixed
= Fix not building properly on Xcode 13 / iOS 15
- Fix problem with timeouts in writeAndAssemble not working as expected if partial data is received- Fix problem with the queue not restarting after transferring states back to Bluejay

## [0.8.8] - 2021-01-20
### Changed
- Replace direct use of XCGLogger with a logging callback interface

### Fixed
- Fix problem with the queue not restarting after transferring states back to Bluejay

## [0.8.7] - 2020-05-11
### Added
- Added support for Swift Package Manager

## [0.8.6] - 2020-02-03
### Changed
- Bumped iOS target to 11 and recommended Xcode to 11.3.1

## [0.8.5] - 2019-05-06
### Changed
- Updated source and dependencies to support Swift 5

## [0.8.4] - 2019-04-05
### Fixed
- Fixed a crash when making a Bluejay call right after an explicit disconnect

## [0.8.3] - 2019-04-03
### Fixed
- Fixed disconnection clean up not working and crashing when it's caused by an unpairing attempt

## [0.8.2] - 2019-02-25
### Added
- Added initial support for did modify services delegation

## [0.8.1] - 2019-01-14
### Fixed
- Fixed and improved Carthage instruction

## [0.8.0] - 2019-01-11
### Added
- XCGLogger, and APIs for logging to a file and monitoring log file changes
- Dittojay demo app as a virtual Bluetooth LE heart rate sensor
  - Also allows testing background state restoration

### Changed
- Migrate to Swift 4.2
- Dropped support for iOS 9
- Removed outdated or soon-to-be-replaced demo projects and documentation
- Redo, clean up, and improve Bluejay demo app to work with Dittojay demo.
- Restrict public access to `Peripheral`

### Fixed
- Background and listen restoration callbacks
- Multiple listen crash by allowing trapping or replacing an existing listen
- Order of queueing when discovering services and characteristics
- Thread-related crashes by adding main thread safety checks to important Bluejay API calls

## [0.7.1] - 2018-12-06
- Add SwiftLint
- Add custom SwiftLint yaml with minor alterations
- Conform all files to SwiftLint suggestions

## [0.7.0] - 2018-11-02
- Add StartMode to better encapsulate CBCentralManager initialization options
- Add WarningOptions to better encapsulate connect warning options
- Fix handling of expected and unexpected disconnections
- Add disconnect handler to handle auto-reconnect
- Remove all cancellation callbacks; use existing failure blocks and look for BluejayError.cancelled instead if you need to differentiate between a failure caused by cancellation versus a failure caused by other errors
- Return Bluejay Peripheral instead of CBPeripheral for connect, disconnect, and RSSI observer callbacks; so that you don't have to import CoreBluetooth in many of your files anymore, and you can also get a "better/more suitable" peripheral object to work with
- Fix a scan crash caused by not clearing the scan request's global timer
- Remove set privileges on all public Bluejay instance variables; they are all supposed to be read only from the get go, but a few were settable and that was dangerous

## [0.6.5] - 2018-10-22
### Added
- Added an option to `scan` to change the threshold for ignoring discoveries based on insignificant changes to RSSI

## [0.6.4] - 2018-08-02
### Added
- Added warnings against using `scan` with `serviceIdentifiers` set to `nil` or empty.

## [0.6.3] - 2018-07-26
### Added
- Add API to check whether a peripheral is listening to a characteristic
- Allow disabling auto-reconnect when using cancelEverything
- Expose auto-reconnect variable
- Update readme, changelog, and documentation

## [0.6.2] - 2018-05-01
### Fixed
- Fix podspec

## [0.6.1] - 2018-05-01
### Fixed
- Fix all Xcode 9.3 warnings

## [0.6.0] - 2018-05-01
### Added
- Add a new API to allow shutting down Bluejay and transfer the CoreBluetooth manager and delegate to another library or code

## [0.5.1] - 2018-04-09
### Fixed
- We weren't using the correct key name to access the listen cache, so listen restoration was broken. This should now be fixed.

## [0.5.0] - 2018-04-09
### Fixed
- There was a subtle true/false reversal mistake that we didn't catch when bringing in the new error enums. This was causing the second connection request to cancel the first ongoing connection request.

## [0.4.9] - 2018-04-09
### Fixed
- Prevent indefinite flush listen
- Add missing cancellation handling for flush listen
- Dedup listen semaphore signals in flush listen
- Add missing semaphore signal for end listen for flush listen
- Use timeout enum in flush listen

## [0.4.8] - 2018-03-16
### Added
- Add timeout to synchronized listen, which allows handling stale or stuck listens when using the run background task.

## [0.4.7] - 2018-03-16
### Added
- Expose peripheral maximum write size API. Useful when EDL is needed and the maximum write size is unknown.

## [0.4.6] - 2018-02-27
### Fixed
- Connecting immediately after a disconnect wasn't possible due to a strict and slightly incorrect double connect protection. This has been fixed now.

## [0.4.5] - 2018-02-26
### Fixed
- Better management of semaphore locks and releases for write and assemble
- Better usage of end listen for write and assemble
- Allow failing write and assemble if bluetooth becomes unavailable or if there's a disconnection after both the read and listen have been setup correctly

## [0.4.4] - 2018-02-09
### Added
- Use `isBluetoothStateUpdateImminent` to check whether the central manager state is unknown or resetting

## [0.4.3] - 2018-02-05
### Fixed

Bluejay/Bluejay/Error.swift
//
//  Error.swift
//  Bluejay
//
//  Created by Jeremy Chiang on 2017-01-03.
//  Copyright © 2017 Steamclock Software. All rights reserved.
//

import Foundation

/// Errors specific to Bluejay.
public enum BluejayError {
    /// Bluetooth is either turned off or unavailable.
    case bluetoothUnavailable
    /// Bluejay does not support another scanning request if Bluejay is still scanning.
    case multipleScanNotSupported
    /// Bluejay does not support another connection request if Bluejay is already connected or still connecting.
    case multipleConnectNotSupported
    /// Bluejay does not support another disconnection request if Bluejay is still disconnecting.
    case multipleDisconnectNotSupported
    /// A connection request in Bluejay has timed out.
    case connectionTimedOut
    /// Bluejay is not connected to a peripheral.
    case notConnected
    /// A Bluetooth service is not found.
    case missingService(ServiceIdentifier)
    /// A Bluetooth characteristic is not found.
    case missingCharacteristic(CharacteristicIdentifier)
    /// A Bluetooth operation is cancelled.
    case cancelled
    /// Bluejay disconnect is called.
    case explicitDisconnect
    /// Bluejay disconnected unexpectedly.
    case unexpectedDisconnect
    /// A disconnection operation is queued.
    case disconnectQueued
    /// An attempt to listen on a characteristic has timed out.
    case listenTimedOut
    /// An attempt to read a characteristic has failed.
    case readFailed
    /// An attempt to write a characteristic has failed.
    case writeFailed
    /// An attempt to read a value from a characteristic has returned no data unexpectedly.
    case missingData
    /// An attempt to read a range of data has failed due to incorrect bounds or an unexpected length.
    case dataOutOfBounds(start: Int, length: Int, count: Int)
    /// An unexpected peripheral is cached and retrieved from CoreBluetooth.
    case unexpectedPeripheral(PeripheralIdentifier)
    /// iOS will not continue scanning in the background if allow duplicates is turned on.
    case allowDuplicatesInBackgroundNotSupported
    /// iOS will not continue scanning in the background if no service identifiers are specified.
    case missingServiceIdentifiersInBackground
    /// Bluejay does not support further Bluetooth operations while a Bluejay background task is still running.
    case backgroundTaskRunning
    /// Bluejay does not support another Bluejay background task when there is already one that is still running.
    case multipleBackgroundTaskNotSupported
    /// Indefinite flush will not exit.
    case indefiniteFlush
    /// Bluejay has stopped.
    case stopped
    /// Bluejay cannot perform certain actions when background restoration is still in progress.
    case backgroundRestorationInProgress
    /// Startup background task has expired during state restoration.
    case startupBackgroundTaskExpired
    /// The original listen declared that duplicated listens are not allowed.
    case multipleListenTrapped
    /// The original listen declared that it can be replaced by a new listen.
    case multipleListenReplaced
    /// An operation expecting a certain amount of data (i.e. writeAndAssemble received more than expected)
    case tooMuchData(expected: Int, received: Data)
}

extension BluejayError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .bluetoothUnavailable:
            return "Bluetooth unavailable"
        case .multipleScanNotSupported:
            return "Multiple scan is not supported"
        case .multipleConnectNotSupported:
            return "Multiple connect is not supported"
        case .multipleDisconnectNotSupported:
            return "Multiple disconnect is not supported"
        case .connectionTimedOut:
            return "Connection timed out"
        case .notConnected:
            return "Not connected to a peripheral"
        case let .missingService(service):
            return "Service not found: \(service.uuid)"
        case let .missingCharacteristic(characteristic):
            return "Characteristic not found: \(characteristic.uuid)"
        case .cancelled:
            return "Cancelled"
        case .explicitDisconnect:
            return "Explicit disconnect"
        case .unexpectedDisconnect:
            return "Unexpected disconnect"
        case .disconnectQueued:
            return "Disconnection is queued"
        case .listenTimedOut:
            return "Listen timed out"
        case .readFailed:
            return "Read failed"
        case .writeFailed:
            return "Write failed"
        case .missingData:
            return "No data from peripheral"
        case let .dataOutOfBounds(start, length, count):
            return "Cannot extract data with a size of \(count) using start: \(start), length: \(length)"
        case let .unexpectedPeripheral(peripheral):
            return "Unexpected peripheral: \(peripheral.uuid)"
        case .allowDuplicatesInBackgroundNotSupported:
            return "Scanning with allow duplicates while in the background is not supported"
        case .missingServiceIdentifiersInBackground:
            return "Scanning without specifying any service identifiers while in the background is not supported"
        case .backgroundTaskRunning:
            return """
            Regular Bluetooth operation is not available when a background task is running. \
            For reading, writing, and listening, please use only the API found in the Synchronized Peripheral \
            provided to you when working inside a background task block
            """
        case .multipleBackgroundTaskNotSupported:
            return "Multiple background task is not supported"
        case .indefiniteFlush:
            return "Flush listen timeout cannot be none or zero"
        case .stopped:
            return "Bluejay stopped"
        case .backgroundRestorationInProgress:
            return "Background restoration is in progress"
        case .startupBackgroundTaskExpired:
            return "Startup background task expired during state restoration"
        case .multipleListenTrapped:
            return "The current listen cannot be installed because an existing listen on the same characteristic is configured to trap"
        case .multipleListenReplaced:
            return "The current listen has been replaced by a newer listen on the same characteristic"
        case .tooMuchData(let expected, let received):
            return "More data than expected was received from the device (expected: \(expected), got: \(received.count))"
        }
    }
}

extension BluejayError: CustomNSError {

    public static var errorDomain: String {
        return "Bluejay"
    }

    public var errorCode: Int {
        switch self {
        case .bluetoothUnavailable: return 1
        case .multipleScanNotSupported: return 2
        case .multipleConnectNotSupported: return 3
        case .multipleDisconnectNotSupported: return 4
        case .connectionTimedOut: return 5
        case .notConnected: return 6
        case .missingService: return 7
        case .missingCharacteristic: return 8
        case .cancelled: return 9
        case .explicitDisconnect: return 10
        case .unexpectedDisconnect: return 11
        case .disconnectQueued: return 12
        case .listenTimedOut: return 13
        case .readFailed: return 14
        case .writeFailed: return 15
        case .missingData: return 16
        case .dataOutOfBounds: return 17
        case .unexpectedPeripheral: return 18
        case .allowDuplicatesInBackgroundNotSupported: return 19
        case .missingServiceIdentifiersInBackground: return 20
        case .backgroundTaskRunning: return 21
        case .multipleBackgroundTaskNotSupported: return 22
        case .indefiniteFlush: return 23
        case .stopped: return 24
        case .backgroundRestorationInProgress: return 25
        case .startupBackgroundTaskExpired: return 26
        case .multipleListenTrapped: return 27
        case .multipleListenReplaced: return 28
        case .tooMuchData: return 29
        }
    }

    public var errorUserInfo: [String: Any] {
        guard let errorDescription = errorDescription else {
            return [:]
        }

        return [
            NSLocalizedDescriptionKey: errorDescription
        ]
    }
}
